pennylane/optimize/momentum.py

```diff
@@ -55,20 +56,44 @@ def apply_grad(self, grad, x):
         Returns:
             array: the new values :math:`x^{(t+1)}`
         """
-
-        grad_flat = _flatten(grad)
-        x_flat = _flatten(x)
+        args_new = list(args)
 
         if self.accumulation is None:
-            self.accumulation = [self._stepsize * g for g in grad_flat]
-        else:
-            self.accumulation = [
-                self.momentum * a + self._stepsize * g for a, g in zip(self.accumulation, grad_flat)
-            ]
+            self.accumulation = [None] * len(args)
+
+        trained_index = 0
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", True):
+                x_flat = _flatten(arg)
+                grad_flat = _flatten(grad[trained_index])
+                trained_index += 1
+
+                self._update_momentum(index, grad_flat)
+
+                x_new_flat = [e - a for a, e in zip(self.accumulation[index], x_flat)]
+
+                args_new[index] = unflatten(x_new_flat, arg)
```

Comment:
I didn't actually realize so many of the optimizers were repeating the gradient step from scratch 🤔 I wonder if we could have them inherit most of this logic from `GradientDescentOptimizer`?

Out of scope for this PR, however.

Reply:
I did think about that.  

Each tends to have its own formula for `x_new_flat`.  We could extra that equation into another class method, then only have to specify `_initialize_momentum`, `_update_momentum` and `_step_formula` in the inherited classes.

Anytime I decided on a new way of doing things, I had to propagate it through seven different files, which got mildly annoying.

Reply:
That seems mildly annoying. I agree that there's probably a better way of rearranging the optimizers making this a bit easier for future development.